### ISSUE TITLE
Looking for a way to modify the local image tag before comparing it to the server tag

### ISSUE THREAD

**Issue by thomazz-nl** [fmartinou/whats-up-docker#56]
Hi there, thanks for making such a helpful piece of software! I've just recently started using Docker and am wondering if/how I can make whats-up-docker compare tag-names that are slightly different. I thought the include regexes would do the trick, but they seem to only match on the tags of the registry, not on the tag-name of the local image. But maybe I'm just using this all in a odd fashion, please correct me if I do. To illustrate the issue:

![afbeelding](https://user-images.githubusercontent.com/10218628/128103869-5494bdf0-21eb-45a7-8ce9-67326980ee5e.png)

For some images (like nginx-unprivileged) I use a compose file, a .env file containing the tag and a Dockerfile to make some slight changes (like adding a specific user UID/GID). Previously I changed the registry name to indicate that I built that image locally, but that would keep WUD from fetching the tags from the remote registry. Now I use the proper registry+image name, but I add the string "-customized" to the tag to be able to identify the modified images. Sadly WUD does not match this and always detects a new version as available.

Hopefully there's a way to change what part of the local tag-name is used when matching it to the remote tag-name (through the json store or webui would be fine too). Diun seems to have some ways to scan [Dockerfiles](https://crazymax.dev/diun/providers/dockerfile/) and [docker-compose.yaml files](https://crazymax.dev/diun/providers/file/), but I don't think that would do the trick when you use environment vars/ARG in them, and a regex would probably be an easier way to solve this.

**Comment by fmartinou** [comment]
Hi, thank you for your support :)

Currently includeTags regex are only used to filter candidate tags found in the registry not to sanitize local tag names.

I think it'd be possible to run includeTags regex on the local tag to sanitize it using capturing groups when performing the comparison to determine whether there's an update.

For example, for the `php-fpm-adminer`, you would set the following includeTags regex:
```
^(\d+\.\d+\.\d+.fpm-alpine).*$
```

So it'd match your local tag `8.0.9.fpm-alpine-customized-adminer` and the remote candidate tag `8.0.9.fpm-alpine`

In the UI, you'd see for the container:
- `local tag = 8.0.9.fpm-alpine-customized-adminer`
- `remote tag = 8.0.9.fpm-alpine`
- `updateAvailable = false`

I don't know if it'd be clear because we'd see 2 different tag names reporting there is no update 🤔.
And it'd certainly cause breaking changes for users who are already using capturing groups in their regex 😨 .

What do you think?

**Comment by thomazz-nl** [comment]
Good points! It would match my needs, but of course I wouldn't want my request to introduce any breaking changes for others. I don't know if there were any use-cases for capturing groups that made people use them and if it is commonly used.

There's no need to break anything though if it's possible to separate concerns: wud.tag.include/exclude can keep doing what they are meant to do (to match remote tags). Perhaps that would be even clearer when they were named something along the line of wud.tag.registry.include/exclude or wud.tag.remote.include/exclude, but that's just me nitpicking, feel free to ignore ;).

To select the local tag used for comparing, a separate (optional) label for containers could be introduced. Something like wud.tag.(custom|local).match?
* If it's absent, use the tag as it does right now (no breakage for others).
* If it's present: use the given regex (no capturing groups required) to perform a match.
    * If there is a match: save original value as data.image.tag.original while saving the match result as data.image.tag.value so it is used from that point on.

About the visual representation, two items could be introduced to the expanded view of each item:
1. the unprocessed tag (as opposed to the _possibly_ processed contents of "Tag" in the image section).
2. the user specified regex (to match a part of the local tag-name).

If the item is collapsed, I do agree that two different tags resulting in "no update" looks a bit odd, although I wouldn't mind much. If the processed "Tag" would be displayed here, the issue would be solved right? Users can expand the item to see the unprocessed tag. Of course this tag could be displayed in the collapsed state too, but it would probably not look clean while not adding much value either.

Thank you for being being understanding and open to suggestions!

**Comment by fmartinou** [comment]
Thanks for the (smart) proposal!

I got a question:

You say:
> If it's present: use the given regex (no capturing groups required) to perform a match

I don't get how I can get the sanitized value from the original without capturing groups?

If I take the previous example with `php-fpm-adminer`, what regex would you write so that it results as:
-  `data.image.tag.original=8.0.9.fpm-alpine-customized-adminer`
-  `data.image.tag.value=8.0.9.fpm-alpine`
?

**Comment by fmartinou** [comment]
Of course... I must be tired 😅 

By the way, do you see any other valid usecases than yours for this feature?
I don'f find any other examples where we'd want to treat the local tag in a different way than the remote tag (because in normal situation the local image is always pulled from its own registry) 🤔 .

I wouldn't add more complexity for most users, just for an edge case.
For example, it wouldn't be appropriate to see in JSON, notifications, UI... the 2 local tags (original and matched) whereas for 99% of users, it's useless.

So, I must be careful that the new `data.image.tag.original`property  is only present/visible for end-users when the new label `wud.tag.local.match` is used...

I'll try to code something on a branch; you'll be the beta-tester 😉 .

**Comment by thomazz-nl** [comment]
Hehe no worries, I can't count the number of times I was puzzled by an issue, while suddenly seeing the obvious when explaining it to someone else :sweat_smile:.
<br>

> By the way, do you see any other valid usecases than yours for this feature?

No, so far I can't seem to think of any :worried:.
<br>
> because in normal situation the local image is always pulled from its own registry

I'm eager to learn how other people deal with their images when keeping them up-to-date. If it's not a normal situation, I have to consider that I might be using Docker the wrong way. The whats-up-docker image seems based on "node:16-alpine", how would you get notified to update your image when there's a new Node release upstream? Maybe whats-up-docker is a bad example as you probably just automated the whole process in a CI/CD pipeline, but would you push every private image you build/adapt (e.g. nginx, php) to a online registry and process it by CI/CD? It seems to me that there's this choice: you either use pre-built images and have version-notifications or you build/modify an existing image and you're on your own for version-notification. I can't help but feel I'm missing something.

>So, I must be careful that the new data.image.tag.originalproperty is only present/visible for end-users when the new label wud.tag.local.match is used...

Good idea, it makes sense to hide those fields when not used.

>I'll try to code something on a branch; you'll be the beta-tester 😉 .

I'll gladly be the guinea pig :hamster:.

p.s. thanks for inspiring me to finally check how I can use emojis in markdown haha :thumbsup:

**Comment by fmartinou** [comment]
Hi,

I think your issue can be resolved thank to [this feature](https://github.com/fmartinou/whats-up-docker/issues/75).
([please find here the user documentation](https://fmartinou.github.io/whats-up-docker/#/configuration/watchers/?id=transform-the-tags-before-performing-the-analysis))

I didn't try myself but if you write a transform function like this one, it should work:
```
wud.tag.transform='^(\d+\.\d+\.\d+.fpm-alpine).*$ => $1'
```
(the function will be applied to both local & remote tag values)

Please keep me informed! :)

**Comment by thomazz-nl** [comment]
Awesome work, but sadly it doesn't seem to be working completely yet. Example:
local image tag: `1.21.3-alpine-customized`
registry image tag: `1.21.3-alpine`
labels in `docker-compose.yaml`:
```yaml
labels:
    wud.tag.include: ^\d+\.\d+\.\d+-alpine(?:-customized)?$$
    wud.tag.transform: ^(\d+\.\d+\.\d+-alpine)(?:-customized)?$$ => $$1
```

Outcome: ![screenshot](https://user-images.githubusercontent.com/10218628/140439740-0e46e934-2dae-45ac-8a20-651d59bb09dd.png)

Desired outcome: the latest tag on the registry is equal to the (transformed) local tag, so the container entry shouldn't be marked as having a new version available.

Or am I missing something obvious?

**Comment by fmartinou** [comment]
You're right!

I just found [another part in the code](https://github.com/fmartinou/whats-up-docker/blob/master/app/model/container.js#L96) where the transform function must be applied as well before comparing.
That's why it's working for most users but not in your specific use case 😞 

I'm going fo fix it.

**Comment by fmartinou** [comment]
Hi,

Could you please give a try to the version `fmartinou/whats-up-docker:fix__56`?
It should fix your last issue.

<img width="589" alt="2021-11-05 11_58_09-ubuntu  En fonction  - Oracle VM VirtualBox" src="https://user-images.githubusercontent.com/16061231/140504577-6938777c-4e87-4972-b112-31c4da2e79ef.png">

**Comment by thomazz-nl** [comment]
Hi there, that fix__56 works like a charm! Thanks a bunch for adding the transform feature 🥳

**Comment by fmartinou** [comment]
Great!
The fix will be part of the next release.

**Comment by fmartinou** [comment]
Available in `5.11.0`
